Jenkinsfile

```diff
@@ -234,6 +250,13 @@ pipeline {
                             post {
                                 always {
                                     junit testResults: "**/*/TEST*${SUREFIRE_REPORT_NAME_SUFFIX}*.xml", keepLongStdio: true
+                                    stash allowEmpty: true, name: itJacocoStashName, includes: '**/*.exec'
+                                }
+
+                                failure {
+                                    zip zipFile: 'test-reports-it.zip', archive: true, glob: "**/*/surefire-reports/**"
+                                    zip zipFile: 'test-errors-it.zip', archive: true, glob: "**/hs_err_*.log"
+                                    stash allowEmpty: true, name: itFlakyTestStashName, includes: '**/FlakyTests.txt'
```

Comment:
Question for my understanding: 

Why do we need to stash these files for IT tests, but not for the other tests? Or in other words, how can I know for which agent I need to stash/unstash things and for which I don't need it?

Is there same way we can make this clearer, e.g. by given the agents different names?

Reply:
At the beginning of the pipeline, we define an agent:

https://github.com/zeebe-io/zeebe/blob/9c7aa2cf9813e10dd7ddba65a8495bc94d7d5b0b/Jenkinsfile#L22-L30

This becomes our main/default agent for all stages, pre/post phases, etc. Everything in the pipeline, unless otherwise specified, runs in the main agent and thus shares the same workspace.

You can execute certain stages/steps in a different agent. To do this, in the stage, you need to explicitly say that you will use a different agent, e.g.:

https://github.com/zeebe-io/zeebe/blob/9c7aa2cf9813e10dd7ddba65a8495bc94d7d5b0b/Jenkinsfile#L183-L191

While you can't associate an agent block to a variable in a Groovy pipeline (unfortunately - I tried! although who knows, maybe I didn't do it well), they do have a name, or at least an ID - the `label` property of the agent. It will start a pod with that label and name, so you can use that to debug/introspect it with k8s. By convention the main agent is just the build name, and the IT one has a suffix of `_it`. I'm happy to take suggestions if we have better ones :slightly_smiling_face: 

Everything in that stage (so any nested stages/steps) will be executed on that agent, which has its own workspace. Workspaces are completely isolated from each other (as are agents), so any modifications done in one are not reflected in the others. This means, building the distribution in the main agent would only build it there, or running flaky tests analysis would only produce a `FlakyTests.txt` file in the agent where you ran the analysis.

Finally, if you have a stage `A` which contains a nested stage `B`, where `A` is executed in the main agent and `B` is executed in a different agent, the `post` phase of `A` will be executed in the same agent as `A`, so it will only have access to that workspace. So if you want to see results from `B` (e.g. `FlakyTests.txt`), you need to copy it over to the workspace of `A`.

So to answer your question of how you know which agent you need to stash/unstash, it depends on what you want to do. Ideally the agents are entirely independent and we don't need to, but as you saw there's some cases where we want to share build artefacts (e.g. the distro, the `FlakyTests.txt` file, etc.).

Hope that helps!

Reply:
I've added a comment which I hope explains a bit better what I just wrote.